src/logicTemplates/dynamicfee.js

```diff
@@ -0,0 +1,180 @@
+     * @returns {string}
+     */
+    getAddress() {
+        return this.address;
+    }
+
+    /**
+     * signDynamicFee returns the main transaction and signed logic needed to complete the transfer.
+     * These should be sent to the fee payer, who can use GetDynamicFeeTransactions
+     * @param {Uint8Array} secretKey: the secret key for building the logic sig
+     * @param {string} genesisHash: the genesisHash to use for the txn
+     * @returns {Object} object containing json of txnbuilder constructor arguments under "txn" and signed logicsig under "lsig"
+     */
+    signDynamicFee(secretKey, genesisHash) {
+        let keys = nacl.keyPairFromSecretKey(secretKey);
+        let from = address.encode(keys.publicKey);
+        let to = this.receiver;
+        let fee = 0;
+        let amount = this.amount;
+        let closeRemainderTo = this.closeRemainder;
+        let firstRound = this.firstValid;
+        let lastRound = this.lastValid;
+        let lease = this.lease;
+        let txn = {
+            "from": from,
+            "to": to,
+            "fee": fee,
+            "amount": amount,
+            "closeRemainderTo": closeRemainderTo,
+            "firstRound": firstRound,
+            "lastRound": lastRound,
+            "genesisHash": genesisHash,
+            "type": "pay",
+            "lease": lease
+        };
+
+        let lsig = new logicSig.LogicSig(new Uint8Array(this.programBytes), undefined);
+        lsig.sign(secretKey);
+        return {"txn": txn, "lsig": lsig};
+    }
+}
+
+/**
+ * getDynamicFeeTransactions creates and signs the secondary dynamic fee transaction, updates
+ * transaction fields, and signs as the fee payer; it returns both
+ * transactions as bytes suitable for sendRaw.
+ * Parameters:
+ * @param {dict} txn - main transaction from payer's signDynamicFee output (a dict of constructor arguments, NOT a transaction.Transaction)
+ * @param {LogicSig} lsig - the signed logic received from the payer's signDynamicFee output
+ * @param {Uint8Array} privateKey - the private key for the account that pays the fee
+ * @param {int} fee - fee per byte for both transactions
+ * @param {int} firstValid - first protocol round on which both transactions will be valid
+ * @param {int} lastValid - last protocol round on which both transactions will be valid
+ *
+ * @throws on invalid lsig
+ */
+function getDynamicFeeTransactions (txn, lsig, privateKey, fee, firstValid, lastValid) {
+    if (!lsig.verify(address.decode(txn.from).publicKey)) {
+        throw new Error("invalid signature");
+    }
+
+    txn.firstRound = firstValid;
+    txn.lastRound = lastValid;
+    txn.fee = fee;
+    if (txn.fee < transaction.ALGORAND_MIN_TX_FEE) {
+        txn.fee = transaction.ALGORAND_MIN_TX_FEE
+    }
+
+    let keys = nacl.keyPairFromSecretKey(privateKey);
+    let from = address.encode(keys.publicKey);
+
+    // must remove lease and re-add using addLease so that fee calculation will match other SDKs
```

Comment:
You should be able to remove all the `txnObj` stuff and just use `txn` after fixing the calculation above.